activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -118,7 +120,11 @@ def commit
         connection.release_savepoint(savepoint_name)
         super
         parent = connection.transaction_manager.current_transaction
-        records.each { |r| parent.add_record(r) }
+        if parent.uncommittable?
+          commit_records
```

Comment:
@jeremy you are 100% right, https://github.com/rails/rails/pull/18426 , I missed that part once I was doing the refactors, see more explanation on my PR.
This will also make the implementation of a TestTransaction easier, as the actual txn should not need to care of copying records to the parent..